tests/test_calculator.hh

```diff
@@ -715,72 +748,120 @@ namespace rascal {
+          for (auto & neigh_swap : neigh_swap_images) {
+            auto & gradients_neigh{gradients_sparse[neigh_swap]};
+            Eigen::Map<Matrix3Xd_RowMaj_t> grad_coeffs_flat(
+                gradients_neigh[key].data(), 3, n_entries_per_key);
+            grad_coeffs_pairs.block(0, result_idx, 3, n_entries_per_key) +=
+                grad_coeffs_flat;
+          }
           result_idx += n_entries_per_key;
         }
       }
       return grad_coeffs_pairs;
     }
 
    private:
-    RepManager & representation;
+    RepCalculator & representation;
     std::shared_ptr<StructureManager> structure_manager;
     Structure_t atomic_structure;
     typename StructureManager::iterator center_it;
 
     inline void advance_center() { ++this->center_it; }
 
     /**
-     * Swap a ClusterRef (i, j) so it refers to (j, i) instead
+     * Swap a ClusterRef<order=2> (i, j) so it refers to (j, i) instead
+     *
+     * @return std::vector of ClusterRefKeys or order 2 (pair keys) of all pairs
+     *         (j, i') where i' is either i or any of its periodic images within
+     *         the cutoff of j. The atom j, on the other hand, must be a real
+     *         atom (not a ghost or periodic image).
      *
      * @todo wouldn't this be better as a member of StructureManager
      *       (viz. AdaptorNeighbourList<whatever>)?
      */
-    PairRef_t swap_pair_ref(const PairRef_t & pair_ref) {
+    std::vector<PairRefKey_t> swap_pair_ref(const PairRef_t & pair_ref) {
+      const double image_pos_tol = 1E-7;
+      auto center_manager{extract_underlying_manager<0>(structure_manager)};
+      auto atomic_structure{center_manager->get_atomic_structure()};
       // Get the atom index to the corresponding atom tag
-      size_t access_index = structure_manager->get_atom_index(pair_ref.back());
+      size_t access_index{structure_manager->get_atom_index(pair_ref.back())};
       auto new_center_it{structure_manager->get_iterator_at(access_index)};
       // Return cluster ref at which the iterator is currently pointing
       auto && new_center{*new_center_it};
+      // Get the position of the original central atom (to find all its periodic
+      // images)
+      size_t i_index{structure_manager->get_atom_index(pair_ref.front())};
+      auto i_position{structure_manager->get_position(i_index)};
+     
       // Iterate until (j,i) is found
+      std::vector<PairRefKey_t> new_pairs;
+      using Positions_t = Eigen::Matrix<double, 3, Eigen::Dynamic>;
       for (auto new_pair : new_center) {
-        if (new_pair.back() == pair_ref.front()) {
-          return new_pair;
+        // go through the neighbors of atom j and wrap them into the cell 
+        // to test if they match atom i
+        Positions_t i_trial_position = new_pair.get_position();
+        Positions_t i_wrapped_position =
+            atomic_structure.wrap_explicit_positions(i_trial_position);
+        if ((i_wrapped_position - i_position).norm() < image_pos_tol) {
```

Comment:
Is it possible to do this with comparing the `get_index()` of the two atoms (to see if they're images of the same atom) rather than comparing wrapped positions?

Reply:
turns out the answer is "yes"